### PR TITLE
Gadget0 failing control loopback test

### PR DESCRIPTION
**Pull request by H2OBrain**
Contents
- Adds test-support for gadget0 for the stm32f429i-disco board

### CODE REVIEW

**Comment by H2OBrain** [comment]
WIKI proposal

how to test usb (linux) :
- install openocd, pyusb, GCC ARM Embedded
- clone libopencm3
- cd libopencm3
- make
- cd tests/gadget-zero
- edit openocd.stm32f4disco.cfg
    - adapt hla_serial to your debugger board (dmesg?) (just comment it out :))
- edit test_gadget0.py
    - uncomment DUT_SERIAL belonging to your board
- make -f Makefile.stm32f429i-disco all flash
- python -m unittest test_gadget0

**Comment by karlp** [comment]
Aim here is to make the loopback test behave the same way as the linux kernel gadget 0 loopback test.  (Late thoughts, this is control req loopback? not data loopback?)

**Comment by H2OBrain** [comment]
It's definitely a control transfer.. (loopback like..)
- send known data to device (store in static buffer)
- read size of known data from device (read from static buffer)
- compare known data with read data

the device fails in the first step with the word-loss error)
looks like this
```
 - buf_out : '5678901234567890123456789012345678901234567'
 - buf_in  : '9012345678901234567890123456789012345676\x00\x00\xa7'
```

File: tests/gadget-zero/main-stm32f429i-disco.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	
+#ifndef USE_USB_INTERRUPT
+	while (1) {
+		usbd_poll(usbd_dev);
+	}
+#else
+	nvic_enable_irq(NVIC_OTG_HS_IRQ);
+	
+	volatile int i = 1;
+	while (i);
```
Comment (fenugrec) [inline tests/gadget-zero/main-stm32f429i-disco.c]: "while (i)" ? Is there something else modifying "i" ?

Comment (H2OBrain) [inline tests/gadget-zero/main-stm32f429i-disco.c]: no, it's just a while (1) asm(""); or maybe even while (1);
i'm never sure what the compiler is going to optimise out in the end

Comment (fenugrec) [inline tests/gadget-zero/main-stm32f429i-disco.c]: I think the standard "while (1) {} " construct is pretty invincible, and compiles always to an endless loop.

**Comment by karlp** [comment]
I've pulled an ~equivalent ctrl loopback test code locally, (and can confirm some of the other bugs) but I've based it more on the linux gadget0/intel compliance code, rather than making up our own new codes for it.  Will close this once it lands.

**Comment by karlp** [comment]
These tests (or equivalent) have merged, along with the fixes for them. thanks for pushing on this.

**Pull request closed without merging**